### PR TITLE
[mempool] add a pre-insert check.

### PR DESCRIPTION
**Pull request by gregorycollins**
We're re-gossiping old transactions in the mempool -- this change eagerly checks them, at the expense of moving the checkpointer more often.

### CODE REVIEW

File: src/Chainweb/Mempool/InMem.hs [hunk 1]
```diff
@@ -222,9 +223,14 @@ insertInMem cfg lock txs0 = do
...
-        -- TODO: other well-formedness checks go here (e.g. ttl check)
-        return $! sizeOK tx
+    preInsertFilter = if runCheck then preInsertFilterReal else return
+    preInsertFilterReal txs = do
+        let chk = _inmemPreInsertCheck cfg
+        let out1 = V.map sizeOK txs
+        out2 <- chk txs
+        let oks = V.zipWith (&&) out1 out2
+        -- keep the good txs only
+        return $! V.map fst $ V.filter snd $ V.zip txs oks
```
Comment (fosskers) [inline src/Chainweb/Mempool/InMem.hs]: My hope is that this fuses correctly.

File: src/Chainweb/Mempool/InMem.hs [hunk 2]
```diff
@@ -203,10 +203,11 @@ maxNumPending = 100000
 ------------------------------------------------------------------------------
 insertInMem :: InMemConfig t    -- ^ in-memory config
             -> MVar (InMemoryMempoolData t)  -- ^ in-memory state
+            -> Bool
```
Comment (fosskers) [inline src/Chainweb/Mempool/InMem.hs]: Boolean blindness?

File: test/Chainweb/Test/Mempool.hs [hunk 3]
```diff
@@ -113,38 +122,69 @@ mempoolTestCase name test (MempoolWithFunc withMempool) =
...
   where
     go = monadicIO (gen >>= run . tout . withMempool . test
                         >>= either fail return)
 
     tout m = timeout 60000000 m >>= maybe (fail "timeout") return
 
-testStartup :: MempoolBackend MockTx -> IO ()
-testStartup = const $ return ()
+testStartup :: InsertCheck -> MempoolBackend MockTx -> IO ()
+testStartup = const $ const $ return ()
```
Comment (fosskers) [inline test/Chainweb/Test/Mempool.hs]: Does `mempty` work here?

File: test/Chainweb/Test/Mempool.hs [hunk 4]
```diff
@@ -113,38 +122,69 @@ mempoolTestCase name test (MempoolWithFunc withMempool) =
...
+    go = runExceptT $ do
+        liftIO $ modifyMVar_ gossipMV $ const $ return checkNotBad
+        liftIO $ insert $ txs ++ badTxs
+        liftIO (lookup txs) >>= V.mapM_ lookupIsPending
+        liftIO (lookup badTxs) >>= V.mapM_ lookupIsMissing
+    txcfg = mempoolTxConfig mempool
+    hash = txHasher txcfg
+    insert v = mempoolInsert mempool True $ V.fromList v
+    lookup = mempoolLookup mempool . V.fromList . map hash
+    checkNotBad xs = return $! V.map (not . (`elem` badTxs)) xs
```
Comment (fosskers) [inline test/Chainweb/Test/Mempool.hs]: Naughtly `elem`? Consider `Set`?

File: test/Chainweb/Test/Mempool/InMem.hs [hunk 5]
```diff
@@ -16,12 +18,19 @@ import Chainweb.Utils (Codec(..))
...
+            $ MempoolWithFunc wf
   where
+    wf f = do
+        mv <- newMVar (\v -> V.mapM (const $ return True) v)
+        let cfg = InMemConfig txcfg mockBlockGasLimit 2048 (checkMv mv)
+        InMem.withInMemoryMempool cfg $ f mv
+
+    checkMv mv xs = do
+        f <- readMVar mv
+        f xs
```
Comment (fosskers) [inline test/Chainweb/Test/Mempool/InMem.hs]: `f <$> readMVar mv`

File: test/Chainweb/Test/Mempool/RestAPI.hs [hunk 6]
```diff
@@ -37,38 +38,39 @@ import Network.X509.SelfSigned
...
     -- allow remoteMp to call the local mempool's getBlock (for testing)
     let remoteMp = remoteMp0 { mempoolGetBlock = \a b c -> mempoolGetBlock inmem a b c }
-    return $! TestServer remoteMp inmem tid
+    return $! TestServer remoteMp inmem checkMv tid
   where
-    server :: MVar (MempoolBackend MockTx) -> MVar ClientEnv -> (IO b -> IO a) -> IO a
-    server inmemMv envMv restore =
+    checkMvFunc mv xs = do
+        f <- readMVar mv
+        f xs
```
Comment (fosskers) [inline test/Chainweb/Test/Mempool/RestAPI.hs]: Here too.

File: test/Chainweb/Test/Mempool/Sync.hs [hunk 7]
```diff
@@ -24,17 +24,26 @@ import Chainweb.Mempool.InMem
...
          ]
   where
+    wf f = do
+        mv <- newMVar (V.mapM (const $ return True))
+        let cfg = InMemConfig txcfg mockBlockGasLimit 2048 (checkMv mv)
+        withInMemoryMempool cfg $ f mv
+
+    checkMv mv xs = do
+        f <- readMVar mv
+        f xs
```
Comment (fosskers) [inline test/Chainweb/Test/Mempool/Sync.hs]: And here.

File: src/Chainweb/Mempool/InMem.hs (hunk 1 above)
Comment (gregorycollins) [reply src/Chainweb/Mempool/InMem.hs]: It almost always does :)

File: src/Chainweb/Mempool/InMem.hs (hunk 2 above)
Comment (gregorycollins) [reply src/Chainweb/Mempool/InMem.hs]: Yes fine. Will fix

File: test/Chainweb/Test/Mempool.hs (hunk 4 above)
Comment (gregorycollins) [reply test/Chainweb/Test/Mempool.hs]: It's a test. O(1) elems

File: test/Chainweb/Test/Mempool/InMem.hs (hunk 5 above)
Comment (gregorycollins) [reply test/Chainweb/Test/Mempool/InMem.hs]: Ehhhhh, check again :)

The closest pointfree I got was `readMVar mv >>= ($ xs)` but that's soup

File: test/Chainweb/Test/Mempool.hs (hunk 3 above)
Comment (gregorycollins) [reply test/Chainweb/Test/Mempool.hs]: What are you suggesting

Comment (fosskers) [reply test/Chainweb/Test/Mempool.hs]: I think `const $ mempty` or just `mempty` will compile.

File: test/Chainweb/Test/Mempool/InMem.hs (hunk 5 above)
Comment (fosskers) [reply test/Chainweb/Test/Mempool/InMem.hs]: Oh my bad, you're right. Personally I'd drink that soup, but not all would.

File: src/Chainweb/Mempool/InMem.hs (hunk 2 above)
Comment (gregorycollins) [reply src/Chainweb/Mempool/InMem.hs]: Fixed

**Pull request merged**